README.md
# libnpmsearch [![npm version](https://img.shields.io/npm/v/libnpmsearch.svg)](https://npm.im/libnpmsearch) [![license](https://img.shields.io/npm/l/libnpmsearch.svg)](https://npm.im/libnpmsearch) [![Travis](https://img.shields.io/travis/npm/libnpmsearch.svg)](https://travis-ci.org/npm/libnpmsearch) [![AppVeyor](https://ci.appveyor.com/api/projects/status/github/zkat/libnpmsearch?svg=true)](https://ci.appveyor.com/project/zkat/libnpmsearch) [![Coverage Status](https://coveralls.io/repos/github/npm/libnpmsearch/badge.svg?branch=latest)](https://coveralls.io/github/npm/libnpmsearch?branch=latest)

[`libnpmsearch`](https://github.com/npm/libnpmsearch) is a Node.js library for
programmatically accessing the npm search endpoint. It does **not** support
legacy search through `/-/all`.

## Example

```js
const search = require('libnpmsearch')

console.log(await search('libnpm'))
=>
[
  {
    name: 'libnpm',
    description: 'programmatic npm API',
    ...etc
  },
  {
    name: 'libnpmsearch',
    description: 'Programmatic API for searching in npm and compatible registries',
    ...etc
  },
  ...more
]
```

## Install

`$ npm install libnpmsearch`

## Table of Contents

* [Example](#example)
* [Install](#install)
* [API](#api)
  * [search opts](#opts)
  * [`search()`](#search)
  * [`search.stream()`](#search-stream)

### API

#### <a name="opts"></a> `opts` for `libnpmsearch` commands

`libnpmsearch` uses [`npm-registry-fetch`](https://npm.im/npm-registry-fetch).
All options are passed through directly to that library, so please refer to [its
own `opts`
documentation](https://www.npmjs.com/package/npm-registry-fetch#fetch-options)
for options that can be passed in.

A couple of options of note for those in a hurry:

* `opts.token` - can be passed in and will be used as the authentication token for the registry. For other ways to pass in auth details, see the n-r-f docs.
* `opts.Promise` - If you pass this in, the Promises returned by `libnpmsearch` commands will use this Promise class instead. For example: `{Promise: require('bluebird')}`

#### <a name="search"></a> `> search(query, [opts]) -> Promise`

`query` must be either a String or an Array of search terms.

If `opts.limit` is provided, it will be sent to the API to constrain the number
of returned results. You may receive more, or fewer results, at the endpoint's
discretion.

The returned Promise resolved to an Array of search results with the following
format:

```js
{
  name: String,
  version: SemverString,
  description: String || null,
  maintainers: [
    {
      username: String,
      email: String
    },
    ...etc
  ] || null,
  keywords: [String] || null,
  date: Date || null
}
```

If `opts.limit` is provided, it will be sent to the API to constrain the number
of returned results. You may receive more, or fewer results, at the endpoint's
discretion.

For streamed results, see [`search.stream`](#search-stream).

##### Example

```javascript
await search('libnpm')
=>
[
  {
    name: 'libnpm',
    description: 'programmatic npm API',
    ...etc
  },
  {
    name: 'libnpmsearch',
    description: 'Programmatic API for searching in npm and compatible registries',
    ...etc
  },
  ...more
]
```

#### <a name="search-stream"></a> `> search.stream(query, [opts]) -> Stream`

`query` must be either a String or an Array of search terms.

If `opts.limit` is provided, it will be sent to the API to constrain the number
of returned results. You may receive more, or fewer results, at the endpoint's
discretion.

The returned Stream emits one entry per search result, with each entry having
the following format:

```js
{
  name: String,
  version: SemverString,
  description: String || null,
  maintainers: [
    {
      username: String,
      email: String
    },
    ...etc
  ] || null,
  keywords: [String] || null,
  date: Date || null
}
```

For getting results in one chunk, see [`search`](#search-stream).

##### Example

```javascript
search.stream('libnpm').on('data', console.log)
=>
// entry 1
{
  name: 'libnpm',
  description: 'programmatic npm API',
  ...etc
}
// entry 2
{
  name: 'libnpmsearch',
  description: 'Programmatic API for searching in npm and compatible registries',
  ...etc
}
// etc
```


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

<a name="1.0.0"></a>
# 1.0.0 (2018-08-27)


### Features

* **api:** got API working ([fe90008](https://github.com/npm/libnpmsearch/commit/fe90008))


test/index.js
'use strict'

const figgyPudding = require('figgy-pudding')
const getStream = require('get-stream')
const qs = require('querystring')
const test = require('tap').test
const tnock = require('./util/tnock.js')

const OPTS = figgyPudding({ registry: {} })({
  registry: 'https://mock.reg/'
})

const REG = OPTS.registry
const search = require('../index.js')

test('basic test', t => {
  const query = qs.stringify({
    text: 'oo',
    size: 20,
    quality: 0.65,
    popularity: 0.98,
    maintenance: 0.5
  })
  tnock(t, REG).get(`/-/v1/search?${query}`).once().reply(200, {
    objects: [
      { package: { name: 'cool', version: '1.0.0' } },
      { package: { name: 'foo', version: '2.0.0' } }
    ]
  })
  return search('oo', OPTS).then(results => {
    t.similar(results, [{
      name: 'cool',
      version: '1.0.0'
    }, {
      name: 'foo',
      version: '2.0.0'
    }], 'got back an array of search results')
  })
})

test('search.stream', t => {
  const query = qs.stringify({
    text: 'oo',
    size: 20,
    quality: 0.65,
    popularity: 0.98,
    maintenance: 0.5
  })
  tnock(t, REG).get(`/-/v1/search?${query}`).once().reply(200, {
    objects: [
      { package: { name: 'cool', version: '1.0.0', date: new Date().toISOString() } },
      { package: { name: 'foo', version: '2.0.0' } }
    ]
  })
  return getStream.array(
    search.stream('oo', OPTS)
  ).then(results => {
    t.similar(results, [{
      name: 'cool',
      version: '1.0.0'
    }, {
      name: 'foo',
      version: '2.0.0'
    }], 'has a stream-based API function with identical results')
  })
})

test('accepts a limit option', t => {
  const query = qs.stringify({
    text: 'oo',
    size: 3,
    quality: 0.65,
    popularity: 0.98,
    maintenance: 0.5
  })
  tnock(t, REG).get(`/-/v1/search?${query}`).once().reply(200, {
    objects: [
      { package: { name: 'cool', version: '1.0.0' } },
      { package: { name: 'cool', version: '1.0.0' } },
      { package: { name: 'cool', version: '1.0.0' } },
      { package: { name: 'cool', version: '1.0.0' } }
    ]
  })
  return search('oo', OPTS.concat({ limit: 3 })).then(results => {
    t.equal(results.length, 4, 'returns more results if endpoint does so')
  })
})

test('accepts quality/mainenance/popularity options', t => {
  const query = qs.stringify({
    text: 'oo',
    size: 20,
    quality: 1,
    popularity: 2,
    maintenance: 3
  })
  tnock(t, REG).get(`/-/v1/search?${query}`).once().reply(200, {
    objects: [
      { package: { name: 'cool', version: '1.0.0' } },
      { package: { name: 'cool', version: '1.0.0' } },
      { package: { name: 'cool', version: '1.0.0' } },
      { package: { name: 'cool', version: '1.0.0' } }
    ]
  })
  return search('oo', OPTS.concat({
    quality: 1,
    popularity: 2,
    maintenance: 3
  })).then(results => {
    t.equal(results.length, 4, 'returns more results if endpoint does so')
  })
})

test('sortBy: quality', t => {
  const query = qs.stringify({
    text: 'oo',
    size: 20,
    quality: 1,
    popularity: 0,
    maintenance: 0
  })
  tnock(t, REG).get(`/-/v1/search?${query}`).once().reply(200, {
    objects: [
      { package: { name: 'cool', version: '1.0.0' } },
      { package: { name: 'cool', version: '1.0.0' } },
      { package: { name: 'cool', version: '1.0.0' } },
      { package: { name: 'cool', version: '1.0.0' } }
    ]
  })
  return search('oo', OPTS.concat({
    sortBy: 'quality'
  })).then(results => {
    t.equal(results.length, 4, 'returns more results if endpoint does so')
  })
})

test('sortBy: popularity', t => {
  const query = qs.stringify({
    text: 'oo',
    size: 20,
    quality: 0,
    popularity: 1,
    maintenance: 0
  })
  tnock(t, REG).get(`/-/v1/search?${query}`).once().reply(200, {
    objects: [
      { package: { name: 'cool', version: '1.0.0' } },
      { package: { name: 'cool', version: '1.0.0' } },
      { package: { name: 'cool', version: '1.0.0' } },
      { package: { name: 'cool', version: '1.0.0' } }
    ]
  })
  return search('oo', OPTS.concat({
    sortBy: 'popularity'
  })).then(results => {
    t.equal(results.length, 4, 'returns more results if endpoint does so')
  })
})

test('sortBy: maintenance', t => {
  const query = qs.stringify({
    text: 'oo',
    size: 20,
    quality: 0,
    popularity: 0,
    maintenance: 1
  })
  tnock(t, REG).get(`/-/v1/search?${query}`).once().reply(200, {
    objects: [
      { package: { name: 'cool', version: '1.0.0' } },
      { package: { name: 'cool', version: '1.0.0' } },
      { package: { name: 'cool', version: '1.0.0' } },
      { package: { name: 'cool', version: '1.0.0' } }
    ]
  })
  return search('oo', OPTS.concat({
    sortBy: 'maintenance'
  })).then(results => {
    t.equal(results.length, 4, 'returns more results if endpoint does so')
  })
})

test('sortBy: optimal', t => {
  const query = qs.stringify({
    text: 'oo',
    size: 20,
    quality: 0.65,
    popularity: 0.98,
    maintenance: 0.5
  })
  tnock(t, REG).get(`/-/v1/search?${query}`).once().reply(200, {
    objects: [
      { package: { name: 'cool', version: '1.0.0' } },
      { package: { name: 'cool', version: '1.0.0' } },
      { package: { name: 'cool', version: '1.0.0' } },
      { package: { name: 'cool', version: '1.0.0' } }
    ]
  })
  return search('oo', OPTS.concat({
    sortBy: 'optimal'
  })).then(results => {
    t.equal(results.length, 4, 'returns more results if endpoint does so')
  })
})

test('detailed format', t => {
  const query = qs.stringify({
    text: 'oo',
    size: 20,
    quality: 0,
    popularity: 0,
    maintenance: 1
  })
  const results = [
    {
      package: { name: 'cool', version: '1.0.0' },
      score: {
        final: 0.9237841281241451,
        detail: {
          quality: 0.9270640902288084,
          popularity: 0.8484861649808381,
          maintenance: 0.9962706951777409
        }
      },
      searchScore: 100000.914
    },
    {
      package: { name: 'ok', version: '2.0.0' },
      score: {
        final: 0.9237841281451,
        detail: {
          quality: 0.9270602288084,
          popularity: 0.8461649808381,
          maintenance: 0.9706951777409
        }
      },
      searchScore: 1000.91
    }
  ]
  tnock(t, REG).get(`/-/v1/search?${query}`).once().reply(200, {
    objects: results
  })
  return search('oo', OPTS.concat({
    sortBy: 'maintenance',
    detailed: true
  })).then(res => {
    t.deepEqual(res, results, 'return full-format results with opts.detailed')
  })
})

test('space-separates and URI-encodes multiple search params', t => {
  const query = qs.stringify({
    text: 'foo bar:baz quux?=',
    size: 1,
    quality: 1,
    popularity: 2,
    maintenance: 3
  })
  tnock(t, REG).get(`/-/v1/search?${query}`).reply(200, { objects: [] })
  return search(['foo', 'bar:baz', 'quux?='], OPTS.concat({
    limit: 1,
    quality: 1,
    popularity: 2,
    maintenance: 3
  })).then(
    () => t.ok(true, 'sent parameters correctly urlencoded')
  )
})


index.js
'use strict'

const figgyPudding = require('figgy-pudding')
const getStream = require('get-stream')
const npmFetch = require('npm-registry-fetch')

const SearchOpts = figgyPudding({
  detailed: { default: false },
  limit: { default: 20 },
  from: { default: 0 },
  quality: { default: 0.65 },
  popularity: { default: 0.98 },
  maintenance: { default: 0.5 },
  sortBy: {}
})

module.exports = search
function search (query, opts) {
  return getStream.array(search.stream(query, opts))
}
search.stream = searchStream
function searchStream (query, opts) {
  opts = SearchOpts(opts)
  switch (opts.sortBy) {
    case 'optimal': {
      opts = opts.concat({
        quality: 0.65,
        popularity: 0.98,
        maintenance: 0.5
      })
      break
    }
    case 'quality': {
      opts = opts.concat({
        quality: 1,
        popularity: 0,
        maintenance: 0
      })
      break
    }
    case 'popularity': {
      opts = opts.concat({
        quality: 0,
        popularity: 1,
        maintenance: 0
      })
      break
    }
    case 'maintenance': {
      opts = opts.concat({
        quality: 0,
        popularity: 0,
        maintenance: 1
      })
      break
    }
  }
  return npmFetch.json.stream('/-/v1/search', 'objects.*',
    opts.concat({
      query: {
        text: Array.isArray(query) ? query.join(' ') : query,
        size: opts.limit,
        quality: opts.quality,
        popularity: opts.popularity,
        maintenance: opts.maintenance
      },
      mapJson (obj) {
        if (obj.package.date) {
          obj.package.date = new Date(obj.package.date)
        }
        if (opts.detailed) {
          return obj
        } else {
          return obj.package
        }
      }
    })
  )
}
